Add e2e tests to svg editor rte

rohitkatlaa opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of NA
2. This PR does the following: It adds the e2e tests for the svgdiagram rte component and also adds two more tools i.e bring forward and send backward which changes the overlapping of objects. It also removes the flag that was hiding svg diagram rte.

Edit: 
10 Screenshots of e2e tests.
[e2e screenshots.zip](https://github.com/oppia/oppia/files/4857321/e2e.screenshots.zip)
5 Circle CI e2e tests screenshots
[circle ci screenshots.zip](https://github.com/oppia/oppia/files/4873055/circle.ci.screenshots.zip)



## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

rohitkatlaa committed:
Merge pull request #17 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #18 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #19 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #20 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #21 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #22 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #23 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #24 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #25 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #26 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #27 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #28 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #29 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #30 from oppia/develop

update

rohitkatlaa committed:
Merge pull request #31 from oppia/develop

Only expose thread data through fetchThreads (#8496)

rohitkatlaa committed:
Merge branch 'develop' of https://github.com/rohitkatlaa/oppia into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

rohitkatlaa committed:
adding e2e tests and other tools

rohitkatlaa committed:
adding protractor file

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
adding tests

rohitkatlaa committed:
completing the e2e tests

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
linting

kevintab95 commented on extensions/objects/templates/svg-filename-editor.component.ts:
```diff
@@ -124,7 +124,7 @@ angular.module('oppia').component('svgFilenameEditor', {
         italic: false
       };
-      ctrl.enableRemoveButton = false;
+      ctrl.isObjectSelected = false;
```
Use objectIsSelected instead. We generally prefix `is` to functions.

rohitkatlaa replied:
Done.

kevintab95 commented on extensions/objects/templates/svg-filename-editor.component.ts:
```diff
@@ -759,10 +767,12 @@ angular.module('oppia').component('svgFilenameEditor', {
             var points = [x, y, x, y];
             var size = ctrl.fabricjsOptions.size;
+            var stroke = ctrl.fabricjsOptions.stroke;
+            stroke = stroke.slice(0, -2) + '1)';
```
Why are we doing this?

rohitkatlaa replied:
When drawing a polygon, the initial lines uses stroke color. So if the alpha is zero you won't see the initial diagram, So Iam forcing it to 1 until the polygon is complete. So once the polygon is complete it retains the actual value.

rohitkatlaa replied:
I will add a comment in the code.

kevintab95 replied:
Sounds good. Thanks!

In general, it's always good to add a comment if you get a comment asking why something is the way it is. See tip [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change):
>Tip: If a reviewer asks questions about the "why" behind something, consider proactively adding a clear comment above the relevant line in your code, since the fact that the reviewer had to ask suggests that at least one developer doesn't understand what is going on from the code alone. Otherwise, you'll probably get a follow-up review comment asking you to leave a code comment anyway :)

rohitkatlaa replied:
Done.

kevintab95 reviewed (changes requested):
Thanks @rohitkatlaa! Few minor comments, PTAL! Also, @rohitkatlaa, the standard procedure we follow when introducing new e2e tests is to post screenshots of at least 10 consecutive runs for the test introduced -- this is to ensure that the new test does not introduce any flakiness.

rohitkatlaa committed:
requested changes

rohitkatlaa commented:
I have compressed the screenshots and attached it in the description.

kevintab95 reviewed (approved):
LGTM, thanks @rohitkatlaa!

kevintab95 commented:
@vojtechjelinek @nithusha21 @U8NWXD @aks681 PTAL!

U8NWXD commented on extensions/rich_text_components/Svgdiagram/protractor.js:
```diff
@@ -0,0 +1,85 @@
+  process.cwd() + '/core/tests/protractor_utils/waitFor.js');
+var request = require('request');
+
+var svgTags = {
```
Since this is a constant, can you make it all-caps?

rohitkatlaa replied:
Done.

U8NWXD commented on extensions/rich_text_components/Svgdiagram/protractor.js:
```diff
@@ -0,0 +1,85 @@
+  'ext></g></svg>')
+};
+
+var customizeComponent = async function(modal, shape) {
```
Can you add a waitFor to make sure the component has loaded before you click?

rohitkatlaa replied:
Done.

U8NWXD commented on extensions/rich_text_components/Svgdiagram/protractor.js:
```diff
@@ -0,0 +1,85 @@
+
+var expectComponentDetailsToMatch = async function(elem, shapeName) {
+  await elem.element(by.css(
+    '.protractor-test-svg-diagram')).getAttribute('src').then(function(src) {
```
Can you avoid the `.then` by using async-await?

rohitkatlaa replied:
Done.

U8NWXD commented on core/tests/protractor_desktop/extensions.js:
```diff
@@ -55,8 +55,33 @@ describe('rich-text components', function() {
       await richTextEditor.addRteComponent('Math', 'abc');
       await richTextEditor.addRteComponent(
         'Video', 'M7lc1UVf-VE', 10, 100, false);
-      // We put these last as otherwise Protractor sometimes fails to scroll to
-      // and click on them.
```
Is this no longer true?

rohitkatlaa replied:
Since I have split the test into two it is separated and the order doesn't matter here.

U8NWXD commented on core/tests/protractor_desktop/extensions.js:
```diff
@@ -88,12 +107,15 @@ describe('rich-text components', function() {
     await users.logout();
   });
 
+
+
```
Can you remove these? I don't think we need 3 blank lines of separation

rohitkatlaa replied:
Done.

U8NWXD reviewed (changes requested):
@rohitkatlaa Thanks for adding these e2e tests! I left some suggestions. Thanks also for running the test locally to make sure it doesn't flake. Can you also make sure that it doesn't flake when it runs on the CI server? Instructions are in the wiki: https://github.com/oppia/oppia/wiki/End-to-End-Tests. To summarize, we're looking for 5 consecutive passes of the extensions suite. I'm sorry for switching this requirement on you after you already tried 10 times locally. We just decided to implement this today since the e2e flakes are getting so bad

rohitkatlaa committed:
requested changes

vojtechjelinek commented on core/domain/rte_component_registry_test.py:
```diff
@@ -174,6 +169,13 @@ def test_rte_components_are_valid(self):
                 % hyphenated_component_id)
             main_html_file = os.path.join(
                 directives_dir, '%s.directive.html' % hyphenated_component_id)
+            if hyphenated_component_id == 'svgdiagram':
+                main_ts_file = os.path.join(
+                    directives_dir, 'oppia-noninteractive-%s.component.ts'
+                    % hyphenated_component_id)
+                main_html_file = os.path.join(
+                    directives_dir, '%s.component.html'
+                    % hyphenated_component_id)
```
Why is this part needed?

rohitkatlaa replied:
Except svgdiagram all are directive files whereas svgdiagram is a component file, so this additional condition is required.

rohitkatlaa replied:
I will add a similar comment in the code.

vojtechjelinek replied:
With the Angular migration will the other also become components? Then maybe use `TODO` here that this will be fixed after the Angular migratin.

rohitkatlaa replied:
Added the todo #9762 .Done.

vojtechjelinek replied:
Thx

vojtechjelinek reviewed (changes requested):
LGTM! Just one question.

rohitkatlaa committed:
adding todo

vojtechjelinek reviewed (approved):
LGTM!

U8NWXD reviewed (changes requested):
e2e tests look great! Once you post the screenshots of the 5 consecutive passes, I'll approve

aks681 reviewed (approved):
Lgtm as codeowner

rohitkatlaa commented:
@kevinlee12 @U8NWXD The extensions suite is running in travis. All file upload related e2e tests are in circle ci and this test is actually uploading a image file and abith might be adding a similar test with file upload in the rte. So I think this test should be moved to the circle ci(because travis doesn't have file upload feature).

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

U8NWXD commented:
@rohitkatlaa I agree. File upload doesn't work in Travis, so yes, we need to move this to CircleCI. Thanks for noticing this! A suite for extensions that require file uploads was introduced by #9678. Can you use that one?

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
change the test from extensions to fileUploadExtensions

nithusha21 commented on assets/constants.ts:
```diff
@@ -5204,7 +5204,6 @@ export = {
   },
 
   "ENABLE_PREREQUISITE_SKILLS": false,
-  "ENABLE_SVG_EDITOR_RTE": false,
```
Are we sure we are ready to enable this?

rohitkatlaa replied:
Once #9763 is merged ,Kevin and I will fully test the editor once again and then we can merge this pr.

rohitkatlaa commented:
@U8NWXD I ran the circle ci tests 5 times consecutively and attached the screenshots in the descriptions.

nithusha21 reviewed (commented):
Overall looks good. Though, are we ready to release this?

U8NWXD reviewed (approved):
lgtm. Thanks @rohitkatlaa!

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
merge conflicts from develop

rohitkatlaa commented:
I had to add the fileUploadExtensions.js file as it was removed in the develop. It is exactly similar to how it was before.

rohitkatlaa committed:
adding fixed width to canvas and using scroll

rohitkatlaa committed:
updating the size on object selection

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
linting

rohitkatlaa committed:
minor bug fixes

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa committed:
merge resolve

nishantwrp commented:
Hi @rohitkatlaa,
Please update the assignee list if somebody else needs to take the next action. 
Thanks

kevintab95 commented:
Just an update, we're going to test this on a server before merging because this PR enables the diagram editor and we want to be sure it is bug free before releasing. Thanks!
/cc @nithusha21

rohitkatlaa committed:
minor bug fix

kevintab95 commented:
@nithusha21 PTAL!

nithusha21 reviewed (approved):
LGTM thanks!

nithusha21 commented:
@kevintab95 can you request changes until testing is complete?

ankita240796 reviewed (changes requested):
Few concerns that I have after testing the feature:

1. Why do we have a default greyish line inside the editor?
<img width="622" alt="Screenshot 2020-07-13 at 2 46 35 PM" src="https://user-images.githubusercontent.com/15226041/87288198-a668e280-c518-11ea-82ce-77a93f26c432.png"> 

2. The default size of 9px for diagrams and pencil tool seems a bit too large, we should be making it similar to other such editors found online.

3. Can we have small tooltips over the features like stroke, fill and bg to make the purpose clear?

rohitkatlaa commented:
The default grey line is actually the scroll bar. In chrome the scrollbar sometimes vanishes and the user won't know that the canvas is scrollable, so I have made the scrollbar visible at all times.
For the 3rd point is just the word "stroke" enough for the tooltip of stroke color selector or is more info required?

kevintab95 commented:
@rohitkatlaa on desktop do we need to show the scrollbar? Seems like we already show the entire canvas in the screenshot.

kevintab95 commented:
Also, just a thought, "stroke" label is already present above the tool so having the same in the tooltip seems redundant. I think what @ankita240796 meant was to have tooltips so that they explain what the tool does.

ankita240796 commented:
> The default grey line is actually the scroll bar. In chrome the scrollbar sometimes vanishes and the user won't know that the canvas is scrollable, so I have made the scrollbar visible at all times.

Ah ok, the design seems a bit weird to me then. Shouldn't it be at the bottom, seems like a partition as of now.

> For the 3rd point is just the word "stroke" enough for the tooltip of stroke color selector or is more info required?

I agree with what @kevintab95 suggested here - more details would be helpful.

rohitkatlaa committed:
requested changes

rohitkatlaa committed:
Merge remote-tracking branch 'upstream/develop' into add-e2e-tests-to-svg-editor-rte

rohitkatlaa commented:
1) I have removed the default scrollbar with a small tweak and it appears only when it is necessary.
2) I have made the default size as 3px. But note that for text alone I am setting the size to 18(because with 3px the text becomes too small).
3) I have added the tooltips.
Stroke - Stroke Color
Fill - Fill Color
Bg - Background Color
![Screenshot from 2020-07-13 19-16-37](https://user-images.githubusercontent.com/42460632/87312929-ddea8580-c53e-11ea-9dc9-791aed0326b7.png)

rohitkatlaa committed:
e2e changes

ankita240796 commented:
> 1. I have removed the default scrollbar with a small tweak and it appears only when it is necessary.
> 2. I have made the default size as 3px. But note that for text alone I am setting the size to 18(because with 3px the text becomes too small).
> 3. I have added the tooltips.
>    Stroke - Stroke Color
>    Fill - Fill Color
>    Bg - Background Color
>    ![Screenshot from 2020-07-13 19-16-37](https://user-images.githubusercontent.com/42460632/87312929-ddea8580-c53e-11ea-9dc9-791aed0326b7.png)

Can we change the color of tooltip to something light, it is somehow blending with the icon a lot?

rohitkatlaa commented:
If it is only about stroke color tooltip then it will blend with whatever color because the color of the icon changes when you change the color.

ankita240796 commented:
> If it is only about stroke color tooltip then it will blend with whatever color because the color of the icon changes when you change the color.

I did not understand - can't we use a grey background for tooltip color only?

rohitkatlaa committed:
change tooltip color

rohitkatlaa commented:
I have changed the color of the tooltip to be grey. 
![Screenshot from 2020-07-14 15-22-02](https://user-images.githubusercontent.com/42460632/87411801-d92cdc80-c5e5-11ea-826f-8d32362e1c7a.png)
However by what I meant earlier is that you still change the color of this icon right as in the below pic. However but I thinnk grey is better because other icons are black anyway.
![Screenshot from 2020-07-14 15-23-34](https://user-images.githubusercontent.com/42460632/87411948-0aa5a800-c5e6-11ea-9fb8-c88d777baf9a.png)

ankita240796 reviewed (approved):
LGTM from design's perspective, Thanks @rohitkatlaa!

kevintab95 commented:
This is going in! Congrats @rohitkatlaa!
